### DISCUSSION TITLE
Using BookLibConnect to only download AAX file

### DISCUSSION THREAD

**Discussion by vkallsen-TD** [audiamus/BookLibConnect#100] [Q&A]
Hi, I've been downloading all of my Audible books with this tool and then convert them with AaxAudioConverter. It may be a moot point, but it seems like this tool downloads all the audiobooks as both a m4b and an aax file. Then when I go into AaxAudioConverter I can break down the aax file into chapter by chapter m4b files that I can download to my phone to play. So I would end up with an m4b whole book file and a folder of chapter by chapter m4b files. Is there a way to have Book Lib Connect only get meet the aax files? I think at this point I will keep the full book m4b file until I convert to a chapter by chapter m4b files but it's a lot of time that gets waisted for something I'm going to eventually delete. Just wanted to verify there's not a button I'm missing to only download as aax or a  better way to use Book Lib Connect and AaxAudioConverter.

**Comment by audiamus** [comment]
Book Lib Connect never downloads .aax, it always downloads .aaxc.

Let me try to explain why it is done this way:

- All Audible audiobooks are internally [Mpeg4](https://en.wikipedia.org/wiki/MP4_file_format) containers which we commonly know as .mp4 if they contain video.
- Audible audiobooks are audio only, of course. The most common audio codec for Mpeg4 is [AAC](https://en.wikipedia.org/wiki/Advanced_Audio_Coding) and that's what Audible is using, too. But they encrypt the audio. Nonetheless, it's still Mpeg4 (.mp4) with AAC audio. But you can't play it directly because of the encryption.
- Audible uses [AES](https://en.wikipedia.org/wiki/Advanced_Encryption_Standard) for the encryption. That makes sense because AES is very wide-spread. All your https internet connections are encrypted this way. Most modern CPUs have hardware acceleration for AES.
- While always employing AES, Audible has developed two different methods for the encryption keys. 
- The older one, which we see as .aax files, used a short key of only 64bits (8 hex digits, the "activation bytes"), which was relatively easy to break, particularly as it remained the same for every book the customer purchased.
- The newer one we only encounter as files with the Audible app for mobile devices. These files are named .aaxc. The encryption key is much more sophisticated here. It's not only longer, but it also comes with a so-called "seed". That's a bit of free text created by a random number generator that is added in front of the data to be encrypted. which leads to different results for each encryption even with the same key. And Audible also assigns a different key to each book. So, the keys for .aaxc files are individual per customer and book. In essence, they are unbreakable.
- The Audible web servers provide an interface ([API](https://en.wikipedia.org/wiki/API)) which the mobile apps use and also the emulators like Book Lib Connect. Members of the Open Source community analysed that API and figured out the essentials. So, Book Lib Connect does it like the Audible App for mobiles and downloads the book in the .aaxc file format. It also downloads the individual seed and key for the book, which Audible calls a "voucher". The voucher is encrypted, too, but Audible provides a token to decrypt the voucher. The token is assigned to your profile when you register Book Lib Connect with your Audible/Amazon account. Sounds a bit complicated and certainly is, but it's the way this API is intended to be used. So basically. the difference between .aax and .aaxc is the complexity of the encryption key. The API wants us to download .aaxc, so we do it.
- As we also get the voucher via the same API, we can decrypt the book. Book Lib Connect assigns this task to the Open Source [AAXClean](https://github.com/Mbucari/AAXClean) library, which is a port of the relevant section of Open Source [FFmpeg](https://ffmpeg.org/) from C to C#. _(C# ist the programming language Book Lib Connect is written in. FFmpeg is written in C.)_
- Book Lib Connect decrypts the voucher and feeds seed and key to AAXClean. AAXClean takes the encrypted AAC audio stream of the .aaxc stream, and runs AES decryption on it. The result is plain audio. Otherwise, the Mpeg4 container, the .aaxc file, will not be touched, in particular, all meta data remains unaltered. 
- Book Lib Connect saves the decrypted book as file. By convention, its file type is set to .m4b. It could also be set to .m4a or even .mp4 without the need to change a single bit in the file. When Mpeg4 became a common format it was almost always video and the .mp4 type was associated with video. Later, it was noted that the AAC audio codec was better than the MP3 codec, and music was also published as Mpeg4, but with the .m4a file type to distinguish it from video (m4a = **M**peg**4** **a**udio). Internally, the Mpeg4 container just doesn't have a video stream. Later again, to distinguish audio books from music, the file type .m4b was invented (m4b = **M**peg**4** (audio) **b**ook). Once more, just a convention, no internal difference to .m4a.
- So, Book Lib Connect downloads .aaxc, decrypts the AAC audio stream and saves the result as .m4b. No conversion involved, no other alteration, just decryption of the audio. By default, Book Lib Connect deletes the .aaxc file after decryption. But you can also keep it, by changing the respective setting.
- .aax, created by Book Lib Connect, ist just a copy of the .m4b file, with a new file type. The only reason this is done - and this is optional - is to provide compatibility with existing AAX Audio Converter. which only allows .aax files (or the older .aa file which are basically encrypted .mp3). AAX Audio Converter assigns all audio manipulation tasks to FFmpeg (see above), AAXClean was not yet around. And FFmpeg recognizes almost any Mpeg4 container, in particular it is fully aware of the Audible file types .aax and .aaxc as a variant. It also automatically detects that the Book Lib Connect .aax file is already decrypted. AAX Audio Converter, however, is not that smart. That's why you need a dummy .aax key, the "activation bytes". Any key will do.
- Something extra will happen with the chapter when Book Lib Connect exports a book but that's beyond the scope of .aax, .aaxc and .m4b files. 
- Now, finally, when AAX Audio Converter creates new .m4b files, it rebuilds all meta data from scratch, replacing Audible-proprietary tags like "narrator" with more common ones. It also rebuilds the chapter data and restructures it if you split the book into chapters and more.

My plan still is to combine AAX Audio Converter and Book Lib Connect into a single application but that involves significant redesign for which I did not find the time in the last year.

**Reply by kryojenik** [reply]
The detail in this discussion is really useful.  It answered a burning question I had on why the 2 apps.  I have only been using Book Lib Connect and not even bothering with AAX Audio Converter since the app I use plays .m4b files.   It wasn't until that last bullet about rebuilding the metadata to less proprietary meta-data that I may consider running all my content through it.

I think a note (or a matrix) in the READMEs for each app that clearly indicates what each app does that the other does not would be nice to see.  Something that indicates which app to use for each use case. what state the files are in after use of each app, etc...
